tests/unit/include/bout/test_traits.cxx

```diff
@@ -0,0 +1,129 @@
+  return {};
+}
+
+template <class T, class U, typename = bout::utils::EnableIfFieldPerp<T, U>>
+auto specific_function_overloads(T, U) -> OnlyForFieldPerp {
+  return {};
+}
+
+template <class T, class U, class ResultType = bout::utils::EnableIfField<T, U>>
+auto example_function(T, U) -> ResultType {
+  return {};
+}
+
+} // namespace
+
+TEST(BoutTraitsTest, EnableIfFieldOverloads) {
+  using namespace bout::utils;
+  static_assert(std::is_same<CorrectForFields, decltype(function_overloads(
+                                                   std::declval<Field2D>(),
+                                                   std::declval<Field3D>()))>::value,
+                "EnableIfField should enable function_overloads for two Fields");
+  static_assert(
+      std::is_same<CorrectForBoutReal,
+                   decltype(function_overloads(std::declval<Field2D>(),
+                                               std::declval<BoutReal>()))>::value,
+      "EnableIfField should disable one overload of function_overloads.\n"
+      "This static_assert should fail if the `function_overloads(T, BoutReal)` template\n"
+      "is commented out");
+  static_assert(
+      std::is_same<OnlyForField2D,
+                   decltype(specific_function_overloads(std::declval<Field2D>(),
+                                                        std::declval<Field2D>()))>::value,
+      "EnableIfField should pick the correct version of specific_function_overloads");
+  static_assert(
+      std::is_same<OnlyForField3D,
+                   decltype(specific_function_overloads(std::declval<Field3D>(),
+                                                        std::declval<Field3D>()))>::value,
+      "EnableIfField should pick the correct version of specific_function_overloads");
+  static_assert(
+      std::is_same<OnlyForFieldPerp,
+                   decltype(specific_function_overloads(
+                       std::declval<FieldPerp>(), std::declval<FieldPerp>()))>::value,
+      "EnableIfField should pick the correct version of specific_function_overloads");
+}
+
+TEST(BoutTraitsTest, EnableIfFieldReturnType) {
+  using namespace bout::utils;
+  static_assert(
+      std::is_same<Field2D, decltype(example_function(std::declval<Field2D>(),
+                                                      std::declval<Field2D>()))>::value,
+      "EnableIfField should return Field2D for two Field2Ds");
+  static_assert(
+      std::is_same<Field3D, decltype(example_function(std::declval<Field2D>(),
+                                                      std::declval<Field3D>()))>::value,
+      "EnableIfField should return Field3D for a Field2D and a Field3D");
+  static_assert(
+      std::is_same<Field3D, decltype(example_function(std::declval<Field3D>(),
+                                                      std::declval<Field2D>()))>::value,
+      "EnableIfField should return Field3D for a Field2D and a Field3D");
```

Comment:
Is there a reason not to check (`Field3D`,`Field3D`)->`Field3D` here? i.e.
```
  static_assert(
      std::is_same<Field3D, decltype(example_function(std::declval<Field3D>(),
                                                      std::declval<Field3D>()))>::value,
      "EnableIfField should return Field3D for a Field3D and a Field3D");
```

Reply:
Nope, I just got distracted writing the other tests! The syntax here is pretty disgusting and it took me awhile to get it right